plugins/org.jkiss.dbeaver.ui.ai/src/org/jkiss/dbeaver/ui/ai/preferences/AIFunctionCategoriesPreferencePage.java

```diff
@@ -0,0 +1,289 @@
+                    treeViewer.setChecked(function, functionEnabled);
+
+                    if (functionEnabled) {
+                        anyFunctionEnabled = true;
+                    } else {
+                        allFunctionsEnabled = false;
+                    }
+                }
+
+                treeViewer.setChecked(category, allFunctionsEnabled);
+                treeViewer.setGrayed(category, anyFunctionEnabled && !allFunctionsEnabled);
+            }
+        } catch (Exception e) {
+            log.error("Failed to update check state", e);
+        }
+    }
+
+
+    @Override
+    public boolean performOk() {
+        try {
+            Set<String> enabledCategories = new HashSet<>();
+            Set<String> enabledFunctions = new HashSet<>();
+
+            for (Map.Entry<String, List<AIFunctionDescriptor>> entry : categoryMap.entrySet()) {
+                String category = entry.getKey();
+                List<AIFunctionDescriptor> functions = entry.getValue();
+
+                boolean allFunctionsChecked = true;
+                boolean anyFunctionChecked = false;
+
+                for (AIFunctionDescriptor function : functions) {
+                    if (treeViewer.getChecked(function)) {
+                        enabledFunctions.add(function.getId());
+                        anyFunctionChecked = true;
+                    } else {
+                        allFunctionsChecked = false;
+                    }
+                }
+
+                if (allFunctionsChecked && anyFunctionChecked) {
+                    enabledCategories.add(category);
+                }
+            }
+
+            AISettings settings = AISettingsManager.getInstance().getSettings();
+            settings.setEnabledFunctionCategories(enabledCategories);
+            settings.setEnabledFunctions(enabledFunctions);
+            AISettingsManager.getInstance().saveSettings(settings);
+
+            return true;
+        } catch (Exception e) {
+            log.error("Failed to save settings", e);
+            return false;
+        }
+    }
+
+
+    @Override
+    protected void performDefaults() {
+        for (Map.Entry<String, List<AIFunctionDescriptor>> entry : categoryMap.entrySet()) {
+            String category = entry.getKey();
+            treeViewer.setChecked(category, false);
+            treeViewer.setGrayed(category, false);
+
+            for (AIFunctionDescriptor function : entry.getValue()) {
+                treeViewer.setChecked(function, false);
+            }
+        }
+        super.performDefaults();
+    }
+
+
+    private class TreeContentProvider implements ITreeContentProvider {
+        @Override
+        public Object[] getElements(Object inputElement) {
+            if (inputElement instanceof Map) {
+                @SuppressWarnings("unchecked")
+                Map<String, List<AIFunctionDescriptor>> map = (Map<String, List<AIFunctionDescriptor>>) inputElement;
```

Comment:
I suggest converting the input model to a view model that is easier to work with here.

As an example, see `com.dbeaver.data.compare.ui.tools.DCCompareTaskWizardPageObjects.Input`